### PR TITLE
feat: add helloworld-deno

### LINKED ISSUE #2443: Add Hello, World for Deno
## Expected Behavior

There is a sample showing how to run Deno.

## Actual Behavior

There is no sample showing how to run Deno.

## Steps to Reproduce the Problem

1. View Samples (https://github.com/knative/docs/tree/master/community/samples/serving)
1. Observer no folder `helloworld-deno`

### PR DESCRIPTION
**Pull request by grant**
Fixes #2443

### CODE REVIEW

File: community/samples/serving/helloworld-deno/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,18 @@
+FROM hayd/alpine-deno:0.38.0
+EXPOSE 8080
+WORKDIR /app
+
```
Comment (mattmoor) [inline community/samples/serving/helloworld-deno/Dockerfile:4]: Remove trailing whitespace:
```suggestion

```

File: community/samples/serving/helloworld-deno/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,18 @@
+FROM hayd/alpine-deno:0.38.0
+EXPOSE 8080
+WORKDIR /app
+ 
+# Prefer not to run as root.
+USER deno
+
```
Comment (mattmoor) [inline community/samples/serving/helloworld-deno/Dockerfile:7]: Remove trailing whitespace:
```suggestion

```

File: community/samples/serving/helloworld-deno/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+EXPOSE 8080
+WORKDIR /app
+ 
+# Prefer not to run as root.
+USER deno
+ 
+# Cache the dependencies as a layer (the following two steps are re-run only when deps.ts is modified).
+COPY deps.ts .
+RUN deno fetch deps.ts
+
```
Comment (mattmoor) [inline community/samples/serving/helloworld-deno/Dockerfile:11]: Remove trailing whitespace:
```suggestion

```

File: community/samples/serving/helloworld-deno/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+COPY deps.ts .
+RUN deno fetch deps.ts
+ 
+# These steps will be re-run upon each file change in your working directory:
+ADD . .
+# Compile the main app so that it doesn't need to be compiled each startup/entry.
+RUN deno fetch main.ts
+
+# Added to ENTRYPOINT of base image.
+CMD ["--allow-net", "main.ts"]
```
Comment (mattmoor) [inline community/samples/serving/helloworld-deno/Dockerfile:18]: Add trailing newline:
```suggestion
CMD ["--allow-net", "main.ts"]

```

File: community/samples/serving/helloworld-deno/README.md [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+   }
+   ```
+
+  1. Create a new file named `Dockerfile` and copy the code block below into it.
+
+    ```docker
+    FROM hayd/alpine-deno:0.38.0
+    EXPOSE 8080
+    WORKDIR /app
+
```
Comment (mattmoor) [inline community/samples/serving/helloworld-deno/README.md:56]: Remove trailing whitespace:
```suggestion

```

File: community/samples/serving/helloworld-deno/README.md [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+  1. Create a new file named `Dockerfile` and copy the code block below into it.
+
+    ```docker
+    FROM hayd/alpine-deno:0.38.0
+    EXPOSE 8080
+    WORKDIR /app
+    
+    # Prefer not to run as root.
+    USER deno
+
```
Comment (mattmoor) [inline community/samples/serving/helloworld-deno/README.md:59]: Remove trailing whitespace:
```suggestion

```

File: community/samples/serving/helloworld-deno/README.md [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    WORKDIR /app
+    
+    # Prefer not to run as root.
+    USER deno
+    
+    # Cache the dependencies as a layer (the following two steps are re-run only when deps.ts is modified).
+    # Ideally fetch deps.ts will download and compile _all_ external files used in main.ts.
+    COPY deps.ts .
+    RUN deno fetch deps.ts
+
```
Comment (mattmoor) [inline community/samples/serving/helloworld-deno/README.md:64]: Remove trailing whitespace:
```suggestion

```

File: community/samples/serving/helloworld-deno/README.md [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+    # Cache the dependencies as a layer (the following two steps are re-run only when deps.ts is modified).
+    # Ideally fetch deps.ts will download and compile _all_ external files used in main.ts.
+    COPY deps.ts .
+    RUN deno fetch deps.ts
+    
+    # These steps will be re-run upon each file change in your working directory:
+    ADD . .
+    # Compile the main app so that it doesn't need to be compiled each startup/entry.
+    RUN deno fetch main.ts
+
```
Comment (mattmoor) [inline community/samples/serving/helloworld-deno/README.md:69]: Remove trailing whitespace:
```suggestion

```

File: community/samples/serving/helloworld-deno/main.ts [hunk 9]
```diff
@@ -0,0 +1,10 @@
+import { serve } from "./deps.ts";
+
```
Comment (mattmoor) [inline community/samples/serving/helloworld-deno/main.ts:2]: Remove trailing whitespace:
```suggestion

```

File: community/samples/serving/helloworld-deno/main.ts [hunk 10]
```diff
@@ -0,0 +1,10 @@
+import { serve } from "./deps.ts";
+ 
+const PORT = 8080;
+const s = serve(`0.0.0.0:${PORT}`);
+const body = new TextEncoder().encode("Hello Deno\n");
+
```
Comment (mattmoor) [inline community/samples/serving/helloworld-deno/main.ts:6]: Remove trailing whitespace:
```suggestion

```

File: community/samples/serving/helloworld-deno/main.ts [hunk 11]
```diff
@@ -0,0 +1,10 @@
+import { serve } from "./deps.ts";
+ 
+const PORT = 8080;
+const s = serve(`0.0.0.0:${PORT}`);
+const body = new TextEncoder().encode("Hello Deno\n");
+ 
+console.log(`Server started on port ${PORT}`);
+for await (const req of s) {
+  req.respond({ body });
+}
```
Comment (mattmoor) [inline community/samples/serving/helloworld-deno/main.ts:10]: Add trailing newline:
```suggestion
}

```

File: community/samples/serving/helloworld-deno/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,17 @@
+FROM hayd/alpine-deno:0.38.0
+EXPOSE 8080
+WORKDIR /app
+# Prefer not to run as root.
+USER deno
+
```
Comment (mattmoor) [inline community/samples/serving/helloworld-deno/Dockerfile:6]: Remove trailing whitespace:
```suggestion

```

File: community/samples/serving/helloworld-deno/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,17 @@
+FROM hayd/alpine-deno:0.38.0
+EXPOSE 8080
+WORKDIR /app
+# Prefer not to run as root.
+USER deno
+ 
+# Cache the dependencies as a layer (the following two steps are re-run only when deps.ts is modified).
+COPY deps.ts .
+RUN deno fetch deps.ts
+
```
Comment (mattmoor) [inline community/samples/serving/helloworld-deno/Dockerfile:10]: Remove trailing whitespace:
```suggestion

```

File: community/samples/serving/helloworld-deno/README.md (hunk 5 above)
Comment (mattmoor) [inline community/samples/serving/helloworld-deno/README.md:56]: Remove trailing whitespace:
```suggestion

```

File: community/samples/serving/helloworld-deno/README.md (hunk 6 above)
Comment (mattmoor) [inline community/samples/serving/helloworld-deno/README.md:59]: Remove trailing whitespace:
```suggestion

```

File: community/samples/serving/helloworld-deno/README.md (hunk 7 above)
Comment (mattmoor) [inline community/samples/serving/helloworld-deno/README.md:64]: Remove trailing whitespace:
```suggestion

```

File: community/samples/serving/helloworld-deno/README.md (hunk 8 above)
Comment (mattmoor) [inline community/samples/serving/helloworld-deno/README.md:69]: Remove trailing whitespace:
```suggestion

```

File: community/samples/serving/helloworld-deno/main.ts (hunk 9 above)
Comment (mattmoor) [inline community/samples/serving/helloworld-deno/main.ts:2]: Remove trailing whitespace:
```suggestion

```

File: community/samples/serving/helloworld-deno/main.ts (hunk 10 above)
Comment (mattmoor) [inline community/samples/serving/helloworld-deno/main.ts:6]: Remove trailing whitespace:
```suggestion

```

File: community/samples/serving/helloworld-deno/main.ts (hunk 11 above)
Comment (mattmoor) [inline community/samples/serving/helloworld-deno/main.ts:10]: Add trailing newline:
```suggestion
}

```

**Comment by grant** [comment]
Cleaned up whitespace.

**Comment by grant** [comment]
Why are the tests stalled?

File: community/samples/serving/helloworld-deno/Dockerfile [hunk 14]
```diff
@@ -0,0 +1,17 @@
+FROM hayd/alpine-deno:0.38.0
+EXPOSE 8080
```
Comment (steren) [inline community/samples/serving/helloworld-deno/Dockerfile:2]: This is not needed

File: community/samples/serving/helloworld-deno/main.ts [hunk 15]
```diff
@@ -0,0 +1,10 @@
+import { serve } from "./deps.ts";
+
+const PORT = 8080;
```
Comment (steren) [inline community/samples/serving/helloworld-deno/main.ts:3]: Do not hardcode  the value of PORT. Use the $PORT environment variable and if not present fallback to an hardcoded number (8080)

File: community/samples/serving/helloworld-deno/Dockerfile [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+WORKDIR /app
+# Prefer not to run as root.
+USER deno
```
Comment (grayside) [inline community/samples/serving/helloworld-deno/Dockerfile:5]: While not running as root is a good practice, we mostly skip this for simplicity unless the container contents need it.

File: community/samples/serving/helloworld-deno/Dockerfile [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+
+# Cache the dependencies as a layer (the following two steps are re-run only when deps.ts is modified).
+COPY deps.ts .
```
Comment (grayside) [inline community/samples/serving/helloworld-deno/Dockerfile:8]: Prefer `./` as destination parameter.

File: community/samples/serving/helloworld-deno/Dockerfile [hunk 18]
```diff
@@ -0,0 +1,17 @@
+FROM hayd/alpine-deno:0.38.0
```
Comment (grayside) [inline community/samples/serving/helloworld-deno/Dockerfile:1]: Should this use `1.0.0-rc2`?
We prefer not to have patch-level versions in the container tag so bugfixes come through on docker pull. However, upstream doesn't provide that, so 🤷

File: community/samples/serving/helloworld-deno/Dockerfile [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+WORKDIR /app
+# Prefer not to run as root.
+USER deno
+
+# Cache the dependencies as a layer (the following two steps are re-run only when deps.ts is modified).
+COPY deps.ts .
+RUN deno fetch deps.ts
+
+# These steps will be re-run upon each file change in your working directory:
+ADD . .
```
Comment (grayside) [inline community/samples/serving/helloworld-deno/Dockerfile:12]: Prefer `COPY . ./`

File: community/samples/serving/helloworld-deno/Dockerfile (hunk 14 above)
Comment (grant) [reply community/samples/serving/helloworld-deno/Dockerfile]: Removed.

File: community/samples/serving/helloworld-deno/main.ts (hunk 15 above)
Comment (grant) [reply community/samples/serving/helloworld-deno/main.ts]: Using an env var.

File: community/samples/serving/helloworld-deno/Dockerfile (hunk 17 above)
Comment (grant) [reply community/samples/serving/helloworld-deno/Dockerfile:7]: Done.

File: community/samples/serving/helloworld-deno/Dockerfile (hunk 16 above)
Comment (grant) [reply community/samples/serving/helloworld-deno/Dockerfile:4]: Done.

File: community/samples/serving/helloworld-deno/Dockerfile (hunk 19 above)
Comment (grant) [reply community/samples/serving/helloworld-deno/Dockerfile:11]: Done.

File: community/samples/serving/helloworld-deno/Dockerfile (hunk 18 above)
Comment (grant) [reply community/samples/serving/helloworld-deno/Dockerfile:1]: I believe the base image is fine as-is.

**Comment by grant** [comment]
Oh thanks Averi!

**Pull request merged**